Site Editor: Offset insertion point in iframe

david-szabo97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Fixes https://github.com/WordPress/gutenberg/issues/28457 https://github.com/WordPress/gutenberg/issues/28885
We moved the editor content into an iframe. It already handles offsetting the popover if a ref is passed. But insertion point calculates the position on its own which doesn't take the iframe offset into account. This change offsets the insertion point in the iframe so it's placed at the right position.

## How has this been tested?
1. Insert `Paragraph` block, fill it up with some random text
2. Insert `Columns` block, choose 50 / 50
3. Move your cursor between the two `Column` block
4. Plus button at the right position

## Screenshots

**Before**

https://user-images.githubusercontent.com/2256104/105689815-853b7280-5efb-11eb-9908-c27e518789f3.mov

**After**

https://user-images.githubusercontent.com/2256104/106460481-b552ba80-6493-11eb-9746-96934873614f.mov


## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Offset insertion point in iframe

david-szabo97 commented on packages/block-editor/src/components/block-list/insertion-point.js:
```diff
@@ -23,6 +23,24 @@ import { isRTL } from '@wordpress/i18n';
 import Inserter from '../inserter';
 import { getBlockDOMNode } from '../../utils/dom';
 
+function offsetIframe( rect, ownerDocument ) {
```
This functions comes from https://github.com/WordPress/gutenberg/blob/e54a9ec2f97e28db3293628e921e40e81cfe5f76/packages/components/src/popover/index.js#L45

Is there a good package to host this function so we can use it in both `components` and `block-editor` package?

mattwiebe replied:
I'm not sure if there's an clear candidate but maybe the `@wordpress/dom` package could be a fit.

jeyip replied:
I took a brief look at Gutenberg packages and agree that `@wordpress/dom` might be the right place

mattwiebe reviewed (approved):
I tested this in the post editor (no iframe) and site editor (iframe) and it works as advertised. We should be able to clean some conditionals up once we use the iframe everywhere, but this does the job for now.

jeyip commented:
## Testing
### Behavior
- [x] Offsets the insertion point in columns correctly

### Browsers
- [x] Firefox
- [x] Edge
- [x] Chrome
- [x] Safari
- [ ] ~IE11~

It looks like the site editor isn't loading again in IE11. It isn't caused by this PR because I can confirm the same issue on the master branch.

<img width="2004" alt="Screen Shot 2021-02-05 at 11 45 43 AM" src="https://user-images.githubusercontent.com/5414230/107081557-bcd1d680-67a7-11eb-9612-0f2f2be0c93b.png">

jeyip commented:
> We should be able to clean some conditionals up once we use the iframe everywhere, but this does the job for now.

I agree -- I think it's fine to move forward with this for now, but could we create an issue and/or leave a comment explaining the `offsetIframe` function and that we plan return to it when the iframe is used everywhere?

jeyip commented:
> We moved the editor content into an iframe. It already handles offsetting the popover if a ref is passed. But insertion point calculates the position on its own which doesn't take the iframe offset into account.

`NB:` This seems strange. I wonder why the insertion point needs to calculate its own position if the editor is already handling positioning of popovers. It feels like this kind of offsetting logic should be handled by a single source of truth.

jeyip reviewed (commented):
+1 from me when we resolve the conversation about DRYing out offsetIframe from the `components` and `block-editor` packages

david-szabo97 committed:
Move offsetIframe into @wordpress/dom

david-szabo97 committed:
Add tests

david-szabo97 commented:
* Moved `offsetIframe` into `@wordpress/dom`
* Added tests. I wonder what you think about these tests. JSDom doesn't have a layout engine so I had to mock `getBoundingClientRect` and `DOMRect`.

david-szabo97 committed:
Update test

youknowriad commented on packages/dom/src/dom.js:
```diff
@@ -1002,3 +1002,29 @@ export function removeInvalidHTML( HTML, schema, inline ) {
+ *
+ * @return {DOMRect} offsetted bounds
+ */
+export function offsetIframe( rect, ownerDocument ) {
```
now that it's a public API, maybe we should find a better name? @ellatrix any thoughts on this?

david-szabo97 replied:
🤔  How about: `offsetBoundsByOwnerFrame`, `offsetRectByOwnerFrame`, `frameToScreenPosition`, `frameToScreenBounds` ?

ellatrix replied:
Why does it need to be public? Can we not offset the iframe in the popover after we get the anchor rect? Why are we using `getAnchorRect` in the first place? That's supposed to be an old API to be deprecated at some point.

david-szabo97 replied:
@ellatrix Moved `offsetIframe` back into the popover package. This time into the `utils`.

jeyip commented on packages/dom/src/test/dom.js:
```diff
@@ -200,6 +201,70 @@ describe( 'DOM', () => {
+			parent.appendChild( iframe );
+			// JSDom doesn't have a layout engine
+			// so we need to mock getBoundingClientRect and DOMRect.
+			iframe.getBoundingClientRect = () => ( {
```
To make it clearer that we're mocking this function, and to fall more in line with idiomatic jest, could we use a jest method like `mockImplementation`? (not sure if that is the exact one)

```javascript
iframe.getBoundingClientRect = jest.mockImplementation(() => ( {
				width: 100,
				height: 100,
				top: iframeTop,
				left: iframeLeft,
			} ));
```

david-szabo97 replied:
Great idea! Wrapped them in `jest.fn( impl )` which is basically a shorthand for `jest.fn().mockImplementation( impl )`.

jeyip commented on packages/dom/src/test/dom.js:
```diff
@@ -200,6 +201,70 @@ describe( 'DOM', () => {
+
+			const child = document.createElement( 'div' );
+			iframe.contentWindow.document.body.appendChild( child );
+			child.getBoundingClientRect = () => ( {
```
https://github.com/WordPress/gutenberg/pull/28630/files#r572190506 Same as above

jeyip reviewed (commented):
+1 when discussions about `dom` method name and jest mock testing methods are resolved.

ellatrix commented:
When did this regress? When I merged the iframe PR, it worked correctly. Would be great to know what changed broke it.

david-szabo97 commented:
> When did this regress? When I merged the iframe PR, it worked correctly. Would be great to know what changed broke it.

I don't think it was a regress. Popover works fine and offsets its position by the iframe position. But `InsertionPointPopover` used custom logic to calculate its position and I think the bug didn't surface when testing the iframe PR.

david-szabo97 committed:
Wrap mock functions with jest.fn

david-szabo97 committed:
Merge branch 'master' into fix/site-editor-inserter-point-vertical-position

youknowriad commented:
The custom logic to calculate position is new and I added it in #27860

david-szabo97 committed:
Move offsetIframe into popover utils

david-szabo97 committed:
Fix import

david-szabo97 commented:
Moved `offsetIframe` into `popover/utils` to avoid making this function public. Also moved tests so at least we now have test coverage.

jeyip reviewed (approved):
lgtm!

vindl commented:
@youknowriad can we include this in 10.0? It looks pretty broken now.

youknowriad commented:
yes, sounds good to me.
